python/pyspark/sql/connect/client/artifact.py

```diff
@@ -0,0 +1,270 @@
+    assert not Path(file_name).is_absolute()
+    assert file_name.endswith(required_suffix)
+    return Artifact(os.path.join(prefix, file_name), storage)
+
+
+class ArtifactManager:
+    """
+    The Artifact Manager is responsible for handling and transferring artifacts from the local
+    client to the server (local/remote).
+
+    Parameters
+    ----------
+    user_id : str, optional
+        User ID.
+    session_id: str
+        An unique identifier of the session which the artifact manager belongs to.
+    channel: grpc.Channel
+        GRPC Channel instance.
+    """
+
+    # Using the midpoint recommendation of 32KiB for chunk size as specified in
+    # https://github.com/grpc/grpc.github.io/issues/371.
+    CHUNK_SIZE: int = 32 * 1024
+
+    def __init__(self, user_id: Optional[str], session_id: str, channel: grpc.Channel):
+        self._user_context = proto.UserContext()
+        if user_id is not None:
+            self._user_context.user_id = user_id
+        self._stub = grpc_lib.SparkConnectServiceStub(channel)
+        self._session_id = session_id
+
+    def _parse_artifacts(self, path_or_uri: str) -> List[Artifact]:
+        # Currently only local files with .jar extension is supported.
+        uri = path_or_uri
+        if urlparse(path_or_uri).scheme == "":  # Is path?
+            uri = Path(path_or_uri).absolute().as_uri()
+        parsed = urlparse(uri)
+        if parsed.scheme == "file":
+            local_path = url2pathname(parsed.path)
+            name = Path(local_path).name
+            if name.endswith(".jar"):
+                artifact = new_jar_artifact(name, LocalFile(local_path))
+            else:
+                raise RuntimeError(f"Unsupported file format: {local_path}")
+            return [artifact]
+        raise RuntimeError(f"Unsupported scheme: {uri}")
+
+    def _create_requests(self, *path: str) -> Iterator[proto.AddArtifactsRequest]:
+        """Separated for the testing purpose."""
+        return self._add_artifacts(chain(*(self._parse_artifacts(p) for p in path)))
+
+    def _retrieve_responses(
+        self, requests: Iterator[proto.AddArtifactsRequest]
+    ) -> proto.AddArtifactsResponse:
+        """Separated for the testing purpose."""
+        return self._stub.AddArtifacts(requests)
+
+    def add_artifacts(self, *path: str) -> None:
+        """
+        Add a single artifact to the session.
+        Currently only local files with .jar extension is supported.
+        """
+        requests: Iterator[proto.AddArtifactsRequest] = self._create_requests(*path)
+        response: proto.AddArtifactsResponse = self._retrieve_responses(requests)
+        summaries: List[proto.AddArtifactsResponse.ArtifactSummary] = []
+
+        for summary in response.artifacts:
+            summaries.append(summary)
+            # TODO(SPARK-42658): Handle responses containing CRC failures.
+
+    def _add_artifacts(self, artifacts: Iterable[Artifact]) -> Iterator[proto.AddArtifactsRequest]:
```

Comment:
I believe I can optimize this further with sort of fetching it while it's sending the actual request but let me leave this as a future improvement.